File: src/main/java/io/github/binout/soccer/interfaces/rest/PlayersResource.java
package io.github.binout.soccer.interfaces.rest;

import io.github.binout.soccer.application.player.GetAllLeaguePlayers;
import io.github.binout.soccer.application.player.GetAllPlayers;
import io.github.binout.soccer.application.player.GetPlayer;
import io.github.binout.soccer.application.player.ReplacePlayer;
import io.github.binout.soccer.domain.player.Player;
import io.github.binout.soccer.interfaces.rest.model.RestLink;
import io.github.binout.soccer.interfaces.rest.model.RestPlayer;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import javax.websocket.server.PathParam;
import java.util.List;
import java.util.stream.Collectors;

@RestController
@RequestMapping("players")
public class PlayersResource {

    @Autowired
    GetAllPlayers getAllPlayers;

    @Autowired
    GetAllLeaguePlayers getAllLeaguePlayers;

    @Autowired
    ReplacePlayer replacePlayer;

    @Autowired
    GetPlayer getPlayer;

    @GetMapping
    public List<RestPlayer> getAll() {
        return getAllPlayers.execute().map(p -> toRestModel(p)).collect(Collectors.toList());
    }

    @GetMapping("league")
    public List<RestPlayer> getAllLeague() {
        return getAllLeaguePlayers.execute().map(p -> toRestModel(p)).collect(Collectors.toList());
    }

    @PutMapping("{name}")
    public ResponseEntity put(@PathParam("name") String name, RestPlayer restPlayer) {
        replacePlayer.execute(name, restPlayer.getEmail(), restPlayer.isPlayerLeague(), restPlayer.isGoalkeeper());
        return ResponseEntity.ok().build();
    }

    @GetMapping("{name}")
    public ResponseEntity get(String name) {
        return getPlayer.execute(name)
                .map(PlayersResource::toRestModel)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    private static RestPlayer toRestModel(Player p) {
        RestPlayer restPlayer = new RestPlayer(p.name());
        restPlayer.setPlayerLeague(p.isPlayerLeague());
        restPlayer.setGoalkeeper(p.isGoalkeeper());
        p.email().ifPresent(restPlayer::setEmail);
        return restPlayer;
    }


    private static RestPlayer toRestModel(String baseUri, Player p) {
        RestPlayer restPlayer = toRestModel(p);
        restPlayer.addLinks(new RestLink(baseUri + "/" + p.name()));
        return restPlayer;
    }

}

File: src/main/kotlin/io/github/binout/soccer/interfaces/rest/PlayersResource.kt
/*
 * Copyright 2016 Benoît Prioux
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.github.binout.soccer.interfaces.rest

import io.github.binout.soccer.application.*
import javax.inject.Inject
import javax.ws.rs.*
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

@Path("rest/players")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
class PlayersResource(
        @Inject private val getAllPlayers: GetAllPlayers,
        @Inject private val getAllLeaguePlayers: GetAllLeaguePlayers,
        @Inject private val replacePlayer: ReplacePlayer,
        @Inject private val getPlayer: GetPlayer) {

    @GET
    fun all() : List<RestPlayer> = getAllPlayers.execute().map { it.toRestModel() }

    @GET
    @Path("league")
    fun allLeague() : List<RestPlayer> = getAllLeaguePlayers.execute().map { it.toRestModel() }

    @PUT
    @Path("{name}")
    fun put(@PathParam("name") name: String, restPlayer: RestPlayer): Response {
        replacePlayer.execute(name, restPlayer.email, restPlayer.isPlayerLeague, restPlayer.isGoalkeeper)
        return Response.ok().build()
    }

    @GET
    @Path("{name}")
    fun get(name: String): Response = getPlayer.execute(name)
            ?.let { Response.ok(it.toRestModel()).build() }
            ?: Response.status(404).build()

}

@Path("rest/players-stats")
class PlayerStatsResource(val getAllPlayerStats: GetAllPlayerStats) {

    @GET
    fun all() : List<RestPlayerStat> = getAllPlayerStats.execute().map { it.toRestModel() }

}

